post_processing_stages/object_classify_tf_stage.cpp

```diff
@@ -0,0 +1,370 @@
+	}
+
+	*found_label_count = result->size();
+	const int padding = 16;
+
+	while (result->size() % padding)
+	{
+		result->emplace_back();
+	}
+
+	return kTfLiteOk;
+}
+
+std::vector<uint8_t> ObjectClassifyTfStage::yuvToRgb(uint8_t *mem, int w, int h, int stride)
+{
+	uint8_t *Y = mem, *U = mem + h * stride, *V = mem + (5 / 4) * h * stride;
+	int dst;
+	std::vector<uint8_t> output(h * w * 3);
+
+	for (int y = 0; y < h; y++)
+	{
+		for (int x = 0; x < w; x++)
+		{
+			Y = mem + y * stride + x;
+			U = mem + (h * stride) + ((int)floor(y / 2)) * (stride / 2) + ((int)floor(x / 2));
+			V = mem + (5 * h * stride) / 4 + ((int)floor(y / 2)) * (stride / 2) + ((int)floor(x / 2));
+			dst = (x + y * w) * 3;
+
+			int YY = *Y, UU = *U, VV = *V;
+
+			output[dst] = std::clamp<int>((YY + (1.402 * (VV - 128))), 0, 255);
+			output[dst + 1] = std::clamp<int>((YY - 0.345 * (UU - 128) - 0.714 * (VV - 128)), 0, 255);
+			output[dst + 2] = std::clamp<int>((YY + 1.771 * (UU - 128)), 0, 255);
+		}
+	}
+
+	return output;
+}
+
+void ObjectClassifyTfStage::getTopResults(uint8_t *prediction, int prediction_size, size_t num_results,
+										  float threshold_high_, float threshold_low_,
+										  std::vector<std::pair<float, int>> *top_results, TfLiteType input_type)
+{
+	// Will contain top N results in ascending order.
+	std::priority_queue<std::pair<float, int>, std::vector<std::pair<float, int>>, std::greater<std::pair<float, int>>>
+		top_result_pq;
+	std::vector<std::pair<float, int>> top_results_old = top_results_;
+	top_results_.clear();
+
+	const long count = prediction_size; // NOLINT(runtime/int)
+	float value = 0.0;
+
+	for (int i = 0; i < count; ++i)
+	{
+		value = prediction[i] / 255.0;
+		// Only add it if it beats the threshold and has a chance at being in
+		// the top N.
+		if (value < threshold_high_)
+		{
+			continue;
+		}
+
+		top_result_pq.push(std::pair<float, int>(value, i));
+
+		for (int i = 0; i < top_results_old.size(); i++)
+		{
+			if (top_results_old[i].second == i)
+			{
+				top_results_old.erase(top_results_old.begin() + i);
+			}
+		}
+
+		// If at capacity, kick the smallest value out.
+		if (top_result_pq.size() > num_results)
+		{
+			top_result_pq.pop();
+		}
+	}
+
+	if (top_result_pq.size() < num_results)
+	{
+		for (int i = 0; i < top_results_old.size(); i++)
+		{
+			value = prediction[(int)top_results_old[i].second] / 255.0;
+			// Only add it if it beats the threshold and has a chance at being in
+			// the top N.
+			if (value < threshold_low_)
+			{
+				continue;
+			}
+
+			top_result_pq.push(std::pair<float, int>(value, i));
+
+			// If at capacity, kick the smallest value out.
+			if (top_result_pq.size() > num_results)
+			{
+				top_result_pq.pop();
+			}
+		}
+	}
+
+	// Copy to output vector and reverse into descending order.
+	while (!top_result_pq.empty())
+	{
+		top_results->push_back(top_result_pq.top());
+		top_result_pq.pop();
+	}
+	std::reverse(top_results->begin(), top_results->end());
+}
```

Comment:
This seems a bit complicated for what (i think) it is doing.  Perhaps just filling a vector with all results (provided the confidence is above the threshold), and calling std::sort() at the end might be better?  The caller then only needs to look at the top N results.

Reply:
I think that sounds good. Then there's a final step of adding back any of the old results that are not already in the new list, but above the lower threshold. (It's a Schmitt trigger kind of thing.)